### PR TITLE
use butil::IOBuf in WriteChunk

### PR DESCRIPTION
**Pull request by wu-hanqing**
Change-Id: If514d166a27da3066649820a140f28ec193b41a1

### What problem does this PR solve?

Issue Number: None

Problem Summary:

### What is changed and how it works?

What's Changed: zeroize memory copy in chunkserver writechunk

How it Works:

Side effects(Breaking backward compatibility? Performance regression?): 

### Check List

- [ ] Relevant documentation/comments is changed or added
- [x] I acknowledge that all my contributions will be made under the project's license

### CODE REVIEW

File: src/fs/ext4_filesystem_impl.cpp [hunk 1]
```diff
@@ -329,6 +329,33 @@ int Ext4FileSystemImpl::Write(int fd,
     return length;
 }
 
+int Ext4FileSystemImpl::Write(int fd,
+                              butil::IOBuf buf,
```
Comment (bai-charisu) [inline src/fs/ext4_filesystem_impl.cpp:333]: const &

File: src/fs/ext4_filesystem_impl.cpp [hunk 2]
```diff
@@ -329,6 +329,33 @@ int Ext4FileSystemImpl::Write(int fd,
...
+                              uint64_t offset,
+                              int length) {
+    int remainLength = length;
+    int relativeOffset = 0;
+    int retryTimes = 0;
+
+    while (remainLength > 0) {
+        ssize_t ret = buf.pcut_into_file_descriptor(fd, offset, length);
```
Comment (bai-charisu) [inline src/fs/ext4_filesystem_impl.cpp:341]: 第三个参数应该是remainLength吧

File: src/fs/ext4_filesystem_impl.h [hunk 3]
```diff
@@ -52,6 +54,7 @@ class Ext4FileSystemImpl : public LocalFileSystem {
     int List(const string& dirPath, vector<std::string>* names) override;
     int Read(int fd, char* buf, uint64_t offset, int length) override;
     int Write(int fd, const char* buf, uint64_t offset, int length) override;
+    int Write(int fd, butil::IOBuf buf, uint64_t offset, int length) override;
```
Comment (bai-charisu) [inline src/fs/ext4_filesystem_impl.h:57]: const &

File: src/fs/local_filesystem.h [hunk 4]
```diff
@@ -159,6 +160,17 @@ class LocalFileSystem {
...
 
+    /**
+     * 向文件指定区域写入数据
+     * @param fd：文件句柄id，通过Open接口获取
+     * @param buf：待写入数据
+     * @param offset：写入区域的起始偏移
+     * @param length：写入数据的长度
+     * @return 返回成功写入的数据长度，失败返回-1
+     */
+    virtual int Write(int fd, butil::IOBuf buf, uint64_t offset,
```
Comment (bai-charisu) [inline src/fs/local_filesystem.h:171]: 同上

File: test/chunkserver/datastore/chunkfilepool_mock_unittest.cpp [hunk 5]
```diff
@@ -161,7 +162,7 @@ TEST_F(CSChunkfilePoolMockTest, PersistEnCodeMetaInfoTest) {
     {
         EXPECT_CALL(*lfs_, Open(poolMetaPath, _))
             .WillOnce(Return(-1));
-        EXPECT_CALL(*lfs_, Write(_, _, _, _))
+        EXPECT_CALL(*lfs_, Write(_, Matcher<const char*>(_), _, _))
```
Comment (bai-charisu) [inline test/chunkserver/datastore/chunkfilepool_mock_unittest.cpp:165]: 为啥这些还是会带char*的参数？

File: src/fs/ext4_filesystem_impl.cpp (hunk 1 above)
Comment (wu-hanqing) [reply src/fs/ext4_filesystem_impl.cpp:333]: 这里是需要把iobuf写入到fd中，会涉及到对iobuf的修改，所以不能用const引用。
这里只拷贝了一次iobuf，开销不大。

File: test/chunkserver/datastore/chunkfilepool_mock_unittest.cpp (hunk 5 above)
Comment (wu-hanqing) [reply test/chunkserver/datastore/chunkfilepool_mock_unittest.cpp:165]: 这里是因为写入操作可能会与快照的metapage有关，更新metapage调用的是Write(_, const char*,....)这个接口。

File: src/fs/ext4_filesystem_impl.cpp (hunk 2 above)
Comment (wu-hanqing) [reply src/fs/ext4_filesystem_impl.cpp]: Done

**Review by bai-charisu — APPROVED** [review]

File: src/chunkserver/datastore/chunkserver_datastore.h [hunk 6]
```diff
@@ -205,11 +208,24 @@ class CSDataStore {
...
     virtual CSErrorCode WriteChunk(ChunkID id,
                                 SequenceNum sn,
-                                const char * buf,
+                                const butil::IOBuf& buf,
                                 off_t offset,
                                 size_t length,
                                 uint32_t* cost,
                                 const std::string & cloneSourceLocation = "");
+
+    // Deprecated, only use for unit & integration test
```
Comment (ilixiaocui) [inline src/chunkserver/datastore/chunkserver_datastore.h:217]: Deprecated？ 现在的测试中还是要用的吧

File: src/fs/ext4_filesystem_impl.cpp [hunk 7]
```diff
@@ -329,6 +329,33 @@ int Ext4FileSystemImpl::Write(int fd,
...
+int Ext4FileSystemImpl::Write(int fd,
+                              butil::IOBuf buf,
+                              uint64_t offset,
+                              int length) {
+    int remainLength = length;
+    int relativeOffset = 0;
+    int retryTimes = 0;
+
+    while (remainLength > 0) {
+        ssize_t ret = buf.pcut_into_file_descriptor(fd, offset, remainLength);
```
Comment (ilixiaocui) [inline src/fs/ext4_filesystem_impl.cpp:341]: 会分批写入文件吗

File: src/fs/ext4_filesystem_impl.h [hunk 8]
```diff
@@ -52,6 +54,7 @@ class Ext4FileSystemImpl : public LocalFileSystem {
     int List(const string& dirPath, vector<std::string>* names) override;
     int Read(int fd, char* buf, uint64_t offset, int length) override;
```
Comment (ilixiaocui) [inline src/fs/ext4_filesystem_impl.h:55]: read这里返回也会从char*拷贝一次到iobuf?

**Review by ilixiaocui — APPROVED** [review]

**Pull request merged**